### PR TITLE
Adding domain to jaeger.thrift transformer

### PR DESCRIPTION
**Pull request by oibe**

### CODE REVIEW

File: model/converter/thrift/jaeger/from_domain_test.go [hunk 1]
```diff
@@ -0,0 +1,147 @@
+package jaeger
```
Comment (black-adder) [inline model/converter/thrift/jaeger/from_domain_test.go]: interesting, why wasn't the license added to this file?

File: model/converter/thrift/jaeger/from_domain_test.go [hunk 2]
```diff
@@ -0,0 +1,147 @@
...
+
+import (
+	"math/rand"
+	"testing"
+	"time"
+
+	"github.com/stretchr/testify/assert"
```
Comment (black-adder) [inline model/converter/thrift/jaeger/from_domain_test.go]: golang imports

non-jaeger github imports

github.com/uber.jaeger

File: model/converter/thrift/jaeger/from_domain_test.go (hunk 1 above)
Comment (oibe) [inline model/converter/thrift/jaeger/from_domain_test.go]: because I suck. :(

Comment (yurishkuro) [inline model/converter/thrift/jaeger/from_domain_test.go]: the script only adds headers to files that are in git, so you should `git add` them first, then run `make fmt`. But the build should fail anyway if you don't have a header.

File: model/converter/thrift/jaeger/from_domain.go [hunk 3]
```diff
@@ -0,0 +1,102 @@
...
+package jaeger
+
+import (
+	"github.com/uber/jaeger/model"
+	"github.com/uber/jaeger/thrift-gen/jaeger"
+)
+
+func FromDomain(spans []*model.Span) []*jaeger.Span {
+	jSpans := make([]*jaeger.Span, len(spans))
+	dToJ := &domainToJaegerTransformer{}
```
Comment (yurishkuro) [inline model/converter/thrift/jaeger/from_domain.go]: don't create a pointer, you can define methods on the struct receiver directly and avoid heap allocations.

File: model/converter/thrift/jaeger/from_domain.go [hunk 4]
```diff
@@ -0,0 +1,102 @@
...
+	return jSpans
+}
+
+func FromDomainSpan(span *model.Span) *jaeger.Span {
+	dToJ := &domainToJaegerTransformer{}
+	return dToJ.transformSpan(span)
+}
+
+type domainToJaegerTransformer struct {
+}
```
Comment (yurishkuro) [inline model/converter/thrift/jaeger/from_domain.go]: {} on the same line would look better, unless go fmt objects

File: model/converter/thrift/jaeger/from_domain_test.go [hunk 5]
```diff
@@ -0,0 +1,147 @@
...
+	for i := 0; i < numSpans; i++ {
+		modelSpan := modelSpans[i]
+		jaegerSpan := jaegerSpans[i]
+
+		assert.Equal(t, jaegerSpan.GetTraceIdLow(), int64(modelSpan.TraceID.Low))
+		assert.Equal(t, jaegerSpan.GetTraceIdHigh(), int64(modelSpan.TraceID.High))
+		assert.Equal(t, jaegerSpan.GetSpanId(), int64(modelSpan.SpanID))
+		assert.Equal(t, jaegerSpan.GetParentSpanId(), int64(modelSpan.ParentSpanID))
+		assert.Equal(t, jaegerSpan.GetOperationName(), modelSpan.OperationName)
+		assert.Equal(t, uint64(jaegerSpan.StartTime), uint64(modelSpan.StartTime.Nanosecond()/1e3))
```
Comment (yurishkuro) [inline model/converter/thrift/jaeger/from_domain_test.go]: s/1e3/1000/

File: model/converter/thrift/jaeger/from_domain_test.go [hunk 6]
```diff
@@ -0,0 +1,147 @@
...
+	References    []model.SpanRef
+	Flags         int32
+	StartTime     time.Time
+	Duration      time.Duration
+	Tags          model.KeyValues
+	Logs          []model.Log
+	Process       model.Process
+}
+
+func generateRandomSpan(options *spanOptions) *model.Span {
```
Comment (yurishkuro) [inline model/converter/thrift/jaeger/from_domain_test.go]: I think we have a flyer in the restrooms that says "don't put logic in tests". Why all these complications?

The best way I would test this converter is to go through all the fixtures in model/converter/thrift/jaeger/fixtures and do the following:
  1. load the domain model JSON into model.Trace
  2. convert to jaeger.thrift spans using your new converter
  3. convert back to domain model using the other converter in this package
  4. do a deep compare of the domain objects (assert.Equals should do the trick)

File: model/converter/thrift/jaeger/from_domain.go [hunk 7]
```diff
@@ -0,0 +1,156 @@
...
+
+func FromDomainSpan(span *model.Span) *jaeger.Span {
+	dToJ := &domainToJaegerTransformer{}
+	return dToJ.transformSpan(span)
+}
+
+type domainToJaegerTransformer struct{}
+
+func (d *domainToJaegerTransformer) keyValueToTag(kv *model.KeyValue) *jaeger.Tag {
```
Comment (yurishkuro) [inline model/converter/thrift/jaeger/from_domain.go]: change receiver from pointer to value

File: model/converter/thrift/jaeger/test_utils.go [hunk 8]
```diff
@@ -0,0 +1,32 @@
...
+	return spans
+}
+
+func loadBatch(t *testing.T, file string) *jaeger.Batch {
+	var batch jaeger.Batch
+	loadJSON(t, file, &batch)
+	return &batch
+}
+
+func loadJSON(t *testing.T, fileName string, i interface{}) {
```
Comment (yurishkuro) [inline model/converter/thrift/jaeger/test_utils.go]: I would assume toDomain tests already know how to load the JSON files

File: model/converter/thrift/jaeger/test_utils.go [hunk 9]
```diff
@@ -0,0 +1,32 @@
+package jaeger
```
Comment (yurishkuro) [inline model/converter/thrift/jaeger/test_utils.go]: `test_utils` file gets compiled into prod. You don't need to move the functions, they will be accessible from another _test file in the same pkg.

File: model/converter/thrift/jaeger/to_domain.go [hunk 10]
```diff
@@ -60,13 +60,15 @@ func (td toDomain) ToDomainSpan(jSpan *jaeger.Span, jProcess *jaeger.Process) *m
...
 	tags := td.getTags(jSpan.Tags)
+	refs := td.getReferences(jSpan.References)
 	return &model.Span{
 		TraceID: model.TraceID{
 			High: uint64(jSpan.TraceIdHigh),
 			Low:  uint64(jSpan.TraceIdLow),
 		},
 		SpanID:        model.SpanID(jSpan.SpanId),
 		OperationName: jSpan.OperationName,
+		References:    refs,
```
Comment (yurishkuro) [inline model/converter/thrift/jaeger/to_domain.go]: nice catch

File: model/converter/thrift/jaeger/from_domain.go [hunk 11]
```diff
@@ -0,0 +1,162 @@
...
+// converts it into a jaeger.Span.  If errors are found
+// during conversion of tags, then error tags are appended.
+func FromDomainSpan(span *model.Span) *jaeger.Span {
+	dToJ := &domainToJaegerTransformer{}
+	return dToJ.transformSpan(span)
+}
+
+type domainToJaegerTransformer struct{}
+
+func (d domainToJaegerTransformer) keyValueToTag(kv *model.KeyValue) *jaeger.Tag {
```
Comment (yurishkuro) [inline model/converter/thrift/jaeger/from_domain.go]: nit: `t` would've been a better receiver name, as in "transformer".

**Pull request merged**